File: Interpreter/InterpreterPattern.py
# coding: utf8

from abc import ABCMeta, abstractmethod


class Expression(object):
    __metaclass__ = ABCMeta

    @abstractmethod
    def interpret(self):
        pass

    @abstractmethod
    def get_priority(self):
        pass


class NumberExpression(Expression):
    cn_to_number = {}
    cn_to_number[u"零"] = 0
    cn_to_number[u"一"] = 1
    cn_to_number[u"二"] = 2
    cn_to_number[u"三"] = 3
    cn_to_number[u"四"] = 4
    cn_to_number[u"五"] = 5
    cn_to_number[u"六"] = 6
    cn_to_number[u"七"] = 7
    cn_to_number[u"八"] = 8
    cn_to_number[u"九"] = 9

    def __init__(self, chinese_number):
        self._cn = chinese_number

    def interpret(self):
        return self.cn_to_number[self._cn]

    def get_priority(self):
        return 1


class UnitExpression(Expression):
    unit_to_number = {}
    unit_to_number[u"十"] = 10 ** 1
    unit_to_number[u"百"] = 10 ** 2
    unit_to_number[u"千"] = 10 ** 3
    unit_to_number[u"万"] = 10 ** 4
    unit_to_number[u"亿"] = 10 ** 8

    def __init__(self, unit, lefts):
        self._unit = unit
        self._lefts = lefts

    def interpret(self):
        base = sum([left.interpret() for left in self._lefts])
        return base * self.unit_to_number[self._unit]

    def get_priority(self):
        return self.unit_to_number[self._unit]


class Context(object):
    def __init__(self, chinese_number):
        self._cn = chinese_number

    def convert(self):
        stack = []

        for char in self._cn:
            if char == u"零":
                continue
            if char in NumberExpression.cn_to_number:
                stack.append(NumberExpression(char))
                continue
            if char in UnitExpression.unit_to_number:
                lefts = []
                for ind in range(len(stack) - 1, -1, -1):
                    if stack[ind].get_priority() > UnitExpression.unit_to_number[char]:
                        break
                    lefts.insert(0, stack.pop(ind))
                stack.append(UnitExpression(char, lefts))
                continue
            raise ValueError("unknown character")
        return sum(item.interpret() for item in stack)


if __name__ == "__main__":
    print(Context(u"三千二百五十亿四千三百万八千一百二十九").convert())

File: Interpreter/InterpreterPattern.java
import java.util.*;

interface Expression {
    int interpret(Map<String, Map<Character, Integer>> context);
    int getPriority(Map<String, Map<Character, Integer>> context);
}

class NumberExpression implements Expression {
    private Character chineseNumber;

    public NumberExpression(Character chineseNumber) {
        this.chineseNumber = chineseNumber;
    }

    public int interpret(Map<String, Map<Character, Integer>> context) {
        return context.get("numbers").get(chineseNumber);
    }

    public int getPriority(Map<String, Map<Character, Integer>> context) {
        return 1;
    }
}

class UnitExpression implements Expression {
    private Character unit;
    private List<Expression> lefts;

    public UnitExpression(Character unit, List<Expression> lefts) {
        this.unit = unit;
        this.lefts = lefts;
    }

    public int interpret(Map<String, Map<Character, Integer>> context) {
        int base = 0;
        for (Expression left: lefts) 
            base += left.interpret(context);

        return base * context.get("units").get(unit);
    }
    
    public int getPriority(Map<String, Map<Character, Integer>> context) {
        return context.get("units").get(unit);
    }
}

class Converter {
    public static final Map<String, Map<Character, Integer>> context = 
        new HashMap<>();

    static {
        Map<Character, Integer> numbers = new HashMap<>();
        numbers.put('零', 0);
        numbers.put('一', 1);
        numbers.put('二', 2);
        numbers.put('三', 3);
        numbers.put('四', 4);
        numbers.put('五', 5);
        numbers.put('六', 6);
        numbers.put('七', 7);
        numbers.put('八', 8);
        numbers.put('九', 9);
        context.put("numbers", numbers);

        Map<Character, Integer> units = new HashMap<>();
        units.put('十', 10);
        units.put('百', 100);
        units.put('千', 1000);
        units.put('万', 10000);
        units.put('亿', 100000000);
        context.put("units", units);
    }

    public int convert(String chineseNumber) {
        Stack<Expression> stack = new Stack<>();
        for (int i=0; i<chineseNumber.length(); ++i) {
            Character character = chineseNumber.charAt(i);
            if ('零' == character)
                continue;

            if (context.get("numbers").containsKey(character)) {
                stack.push(new NumberExpression(character));
                continue;
            }

            if (context.get("units").containsKey(character)) {
                List<Expression> lefts = new ArrayList<>();

                while (stack.size() > 0 && stack.peek().getPriority(context) 
                    <= context.get("units").get(character))
                    lefts.add(stack.pop());

                stack.push(new UnitExpression(character, lefts));
                continue;
            }
        }

        int sum = 0;
        for (Expression expression: stack)
            sum += expression.interpret(context);
        return sum;
    }
}

class InterpreterPattern {
    public static void main(String[] args) {
        System.out.println(new Converter().convert("五亿三十三万三千三百三十三"));
    }
}

